Add initial type hints and run mypy in CI

intgr opened the pull request:
Closes #337.

intgr commented on rest_framework_nested/runtests/runcoverage.py:
```diff
@@ -34,7 +36,7 @@ def main():
             DeprecationWarning
         )
-        failures = TestRunner(['tests'])
+        failures = TestRunner(['tests'])  # type: ignore[assignment,arg-type]
```
I don't think this code path even works. `TestRunner` doesn't seem to accept list argument.

alanjds replied:
Me neither.

intgr commented on rest_framework_nested/relations.py:
```diff
@@ -81,14 +81,14 @@ def get_object(self, view_name, view_args, view_kwargs):
             return super().to_internal_value(data)
         except ValidationError as err:
-            if err.detail[0].code != 'no_match':
+            if err.detail[0].code != 'no_match':  # type: ignore[union-attr,index]
```
This code is a bit iffy. `err.detail` isn't guaranteed to be a list of `ErrorDetail`. It can be an arbitrarily nested combination of `list`, `dict` and `ErrorDetail`.

intgr commented on rest_framework_nested/relations.py:
```diff
@@ -6,29 +6,29 @@
 from rest_framework.exceptions import ValidationError
+from rest_framework.request import Request
 
-# fix for basestring
-try:
-    basestring
-except NameError:
-    basestring = str
```
Looked like some obsolete Python 2.x compatibility junk?

intgr commented on rest_framework_nested/runtests/urls.py:
```diff
@@ -1,6 +1,6 @@
 Blank URLConf just to keep runtests.py happy.
 """
-from rest_framework.compat import patterns
+from rest_framework.compat import patterns  # type: ignore[attr-defined]
```
`rest_framework.compat.patterns` doesn't actually exist, this crashes with `ImportError`.

intgr commented:
Marking as Draft until #340 is merged.

alanjds reviewed (commented):
Overall, I would try to keep typehints that enhance the legibility, not that enhance automated type checking. However, some compromise is possible in this front.

Instead of using stuff `from typing` I would `from __future__ import annotations` and go with `str | None` and `list` in opposition of `Union[str, None]` and `List[Any]`, for example.

Also, some parts _for me_ confuses the reader more than helps them, like generics and so.

May be not the time to introduce Mypy as a linter. I am more leaned in typehints _for humans_.
I can contribute to this PR if you like, but am tight on time this last weeks.
Over and all, I really appreciate your the contribution.

intgr commented:
> Instead of using stuff from typing I would from __future__ import annotations and go with str | None and list in opposition of Union[str, None] and List[Any], for example.

Indeed, I hate the `Union[]` things. I didn't realize that is supported on relevant Python versions. Will change.

intgr commented:
> Also, some parts for me confuses the reader more than helps them, like generics and so.

I think generics are useful for human understanding too.

For example this tells the story "each `NestedHyperlinkedRelatedField` instance is bound to a certain model type (T_Model). The `get_object()` and `to_internal_value()` methods always return values of this declared model type (not some arbitrary model) and that also applies to inherited methods and attributes, e.g. `RelatedField.get_queryset()`"

![image](https://github.com/alanjds/drf-nested-routers/assets/137616/ee72ba13-1486-421b-8310-8c76db69b2a9)

> May be not the time to introduce Mypy as a linter. I am more leaned in typehints for humans.

I can remove mypy linting from this repo. It was useful for double-checking that my added hints are correct. Though seems like a loss from my PoV.

But some 3rd party projects that use `drf-nested-routers` will want to use type checkers on their project. That was my motivation for adding type hints here in the first place, I use mypy in my own project. I would urge not to remove anything that would degrade typechecking accuracy for downstream users.

alanjds commented on rest_framework_nested/routers.py:
```diff
@@ -69,19 +82,21 @@ def __init__(self, parent_router, parent_prefix, *args, **kwargs):
-        parent_registry = [registered for registered
-                           in self.parent_router.registry
-                           if registered[0] == self.parent_prefix]
+        parent_registry: list[tuple[str, type[ViewSetMixin], str]] = [
```
This comes from upstream code. What about signaling just what we know to be true?
```suggestion
        parent_registry: list[tuple] = [
```

My concern is that upstream changes and this breaks our type checks.

intgr replied:
> My concern is that upstream changes and this breaks our type checks.

Well then it would break the code as well. And it would be good to know when this code needs to be updated.

The tuple gets unpacked a few lines below, you're already very much dependent on these types:

```python
parent_registry_item = parent_registry[0]
parent_prefix, parent_viewset, parent_basename = parent_registry_item
```

intgr replied:
Actually never mind, I can remove this whole hint and mypy will infer it.

alanjds commented on mypy.ini:
```diff
@@ -0,0 +1,14 @@
+[mypy]
+check_untyped_defs = true
+disallow_incomplete_defs = true
+disallow_untyped_defs = true
```
I am ok to try this, and changing them to `false` if future PRs start to not pass because of this.

Is there a way to change this to warnings instead of breaking?
...or the tool can be an "optional" on the list of checks, idk.

intgr replied:
> I am ok to try this, and changing them to false if future PRs start to not pass because of this.

These particular settings cause mypy to fail when methods/functions are added with missing type hints.

The "correct" course of action would be to add the requested hints, instead of ignoring or disabling. :)

In any case, if you get problems from mypy in the future, feel free to ping me in the PR and I will be happy to assist.

alanjds commented on rest_framework_nested/runtests/settings.py:
```diff
@@ -83,7 +85,7 @@
 )
 
-INSTALLED_APPS = (
+INSTALLED_APPS: tuple[str, ...] = (
```
That surprised me, I confess. The tools are not able to infer this one?

intgr replied:
Mypy prefers to infer tuples as non-variadic, e.g. `tuple[str, str, str, str, str, str]` etc.

This causes problems since later lines add values to the tuple `INSTALLED_APPS += (`, changing its length and making it incompatible.

Changing it to a `list` is one possible alternative.

intgr replied:
Replaced them with list literal `[]` syntax

alanjds commented on rest_framework_nested/relations.py:
```diff
@@ -4,24 +4,31 @@
-class NestedHyperlinkedRelatedField(rest_framework.relations.HyperlinkedRelatedField):
+T_Model = TypeVar('T_Model', bound=Model)
+
+
+class NestedHyperlinkedRelatedField(HyperlinkedRelatedField, Generic[T_Model]):
     lookup_field = 'pk'
     parent_lookup_kwargs = {
         'parent_pk': 'parent__pk'
     }
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
         self.parent_lookup_kwargs = kwargs.pop('parent_lookup_kwargs', self.parent_lookup_kwargs)
         super().__init__(*args, **kwargs)
 
-    def get_url(self, obj, view_name, request, format):
+    def get_url(self, obj: Model, view_name: str, request: Request, format: str | None) -> str | None:
```
Idk if `NestedHyperlinkedRelatedField` is always about a `django.db.models.Model`.

intgr replied:
Yes, all `RelatedField`s are about models.

https://www.django-rest-framework.org/api-guide/relations/#serializer-relations

> Relational fields are used to represent model relationships. They can be applied to ForeignKey, ManyToManyField and OneToOneField relationships, as well as to reverse relationships, and custom relationships such as GenericForeignKey.

intgr replied:
btw I copied many of the signatures from `djangorestframework-stubs` where I'm one of the maintainers, it covers most of the API surface of DRF itself.

https://github.com/typeddjango/djangorestframework-stubs/blob/master/rest_framework-stubs/relations.pyi#L133

alanjds commented on rest_framework_nested/viewsets.py:
```diff
@@ -1,24 +1,34 @@
 
 @contextlib.contextmanager
-def _force_mutable(querydict: dict) -> dict:
+def _force_mutable(querydict: QueryDict | dict[str, Any]) -> Generator[QueryDict | dict[str, Any], None, None]:
```
What about reducing this one?
```suggestion
def _force_mutable(querydict: QueryDict | dict[str, Any]) -> Iterable[QueryDict | dict[str, Any]]:
```

From https://docs.python.org/3/library/typing.html#typing.Generator :

> ```
> If your generator will only yield values, set the SendType and ReturnType to None:
> 
> def infinite_stream(start: int) -> Generator[int, None, None]:
>     while True:
>         yield start
>         start += 1
> 
> Alternatively, annotate your generator as having a return type of either Iterable[YieldType] or Iterator[YieldType]:
> 
> def infinite_stream(start: int) -> Iterator[int]:
>     while True:
>         yield start
>         start += 1
> ```

alanjds commented on rest_framework_nested/viewsets.py:
```diff
@@ -29,7 +39,7 @@ def _get_parent_lookup_kwargs(self) -> dict:
 
         if not parent_lookup_kwargs:
-            serializer_class = self.get_serializer_class()
+            serializer_class = cast(GenericAPIView, self).get_serializer_class()
```
This one I am still trying to digest. Seems as a source of confusion to the end developer.

Instead, does typing the `serializer_class` directly achieves the same result?
```suggestion
            serializer_class: Serializer = self.get_serializer_class()
```

intgr replied:
This is a general issue with mixin classes.

This code here in `NestedViewSetMixin` code assumes that whatever code using it always combines it with something that derives from `GenericAPIView`. But it's not actually part of the class hierarchy.

Adding the type hint to `serializer_class` doesn't solve the error from mypy, because the type error is about being unable to resolve `get_serializer_class()` method on `self`.

alanjds commented on rest_framework_nested/viewsets.py:
```diff
@@ -39,28 +49,28 @@ def _get_parent_lookup_kwargs(self) -> dict:
         `serializer_class.parent_lookup_kwargs` or `viewset.parent_lookup_kwargs`
         """
-        queryset = super().get_queryset()
+        queryset = super().get_queryset()  # type: ignore[misc]
```
Is this not acceptable?

```suggestion
        queryset: QuerySet = super().get_queryset()  # type: ignore[misc]
```

intgr replied:
The mypy error is "error: "get_queryset" undefined in superclass", because this is a mixin class and mypy doesn't know what the bases are

Elsewhere I have solved this with a cast

```suggestion
        queryset = cast(ModelViewSet, super()).get_queryset()
```

alanjds replied:
I see. In this case, I vote for `# type: ignore[misc]` everywhere the `cast()` appeared.

alanjds commented on rest_framework_nested/viewsets.py:
```diff
@@ -39,28 +49,28 @@ def _get_parent_lookup_kwargs(self) -> dict:
         parent_lookup_kwargs = self._get_parent_lookup_kwargs()
         for query_param, field_name in parent_lookup_kwargs.items():
-            orm_filters[field_name] = self.kwargs[query_param]
+            orm_filters[field_name] = cast(ViewSetMixin, self).kwargs[query_param]
```
Again, is it ok to type only the result instead of `cast` ing the source?

alanjds commented on rest_framework_nested/viewsets.py:
```diff
@@ -39,28 +47,28 @@ def _get_parent_lookup_kwargs(self) -> dict:
         """
-        request = super().initialize_request(request, *args, **kwargs)
-        
+        request = cast(ViewSetMixin, super()).initialize_request(request, *args, **kwargs)
```
My concern on this is about the newer devs needing to reason that.

alanjds commented on tox.ini:
```diff
@@ -10,7 +10,9 @@ python =
 
 [testenv]
-commands = ./runtests.py --nolint
+commands =
+       ./runtests.py --nolint
+       mypy rest_framework_nested
```
If possible, please let `mypy` be another check command, that can fail separately from the runtests one.

alanjds reviewed (commented):
For me, this PR is way more readable now. Am still concerned about some verbosity added but there is no much we can do about this iiuc.

Specifically about the `cast`s, I would prefer to not include them, and mark only the result types instead.

Specifically about `mypy`, I am in favor of letting it running in a non-mandatory way for PR merging. Or at least in a way very easy to turn to non-mandatory in the future.

intgr commented:
> Specifically about the casts, I would prefer to not include them, and mark only the result types instead.

I replied in 2 of the threads, that doesn't solve the issue.

There are a few options: using `cast()` or a `# type: ignore` comment or disabling mypy entirely. (I also tried duplicating the signatures of DRF's methods, but that created more issues than it solved)

> Specifically about mypy, I am in favor of letting it running in a non-mandatory way for PR merging. Or at least in a way very easy to turn to non-mandatory in the future.

How to implement it as "non-mandatory"? Just a separate GitHub job, but still reporting the success/error status? Or ignoring its status entirely and always reporting success?

alanjds commented:
> There are a few options: using `cast()` or a `# type: ignore` comment or disabling mypy entirely. (I also tried duplicating the signatures of DRF's methods, but that created more issues than it solved)

I see :/ . Well, I prefer to go with `# type: ignore`  for now, as it does not confuse/clutter the code.

> > Specifically about mypy,...
> 
> How to implement it as "non-mandatory"? Just a separate GitHub job, but still reporting the success/error status? Or ignoring its status entirely and always reporting success?

A separate GitHub job is better, as is just a GitHub config to let it mandatory or not. This way, the success/error status can be reported correctly.

intgr commented on rest_framework_nested/routers.py:
```diff
@@ -45,7 +49,16 @@ class LookupMixin:
 
 class NestedMixin:
-    def __init__(self, parent_router, parent_prefix, *args, **kwargs):
+    trailing_slash: str
+    routes: list[Route | DynamicRoute]
```
Sure, it means adding a few more `type: ignore` comments.

intgr commented:
I have split mypy into a separate job and removed some hints according to feedback.

If what you mean by "non-mandatory" is jobs without these "Required" labels, then this is actually part of the repository admin confguration in GitHub web UI, not part of the workflow files. Jobs are "non-required" by default unless explicitly added there. So that part should be solved.

> ![image](https://github.com/alanjds/drf-nested-routers/assets/137616/bbf4db76-85bc-485c-9e4f-1877acb608ca)

intgr commented on rest_framework_nested/viewsets.py:
```diff
@@ -39,35 +48,35 @@ def _get_parent_lookup_kwargs(self) -> dict:
         """
-        request = super().initialize_request(request, *args, **kwargs)
-        
+        drf_request: Request = super().initialize_request(request, *args, **kwargs)  # type: ignore[misc]
```
I needed to rename this variable, unfortunately, because it clashes with parameter name `request: HttpRequest`.

Using `cast()`, mypy allows reassigning a variable with a different type, but not when using an explicit variable hint `request: Request = ...`

alanjds commented on requirements-tox.txt:
```diff
@@ -6,6 +6,11 @@ flake8==6.1.0
 ipdb==0.13.13
 pytz==2024.1
 
+# Type checking requirements
+mypy==1.10.0
+django-stubs==5.0.0
+djangorestframework-stubs==3.15.0
```
I am fine with any of the options, yet less files seems preferable.

alanjds commented:
That is great. Thanks a lot for the effort @intgr and the patience on the code review roundtrips.

intgr commented:
Thanks! Before this is usable for downstreams, we still need to add `py.typed` file, so don't release this yet. 😄

alanjds commented:
FYI: Is just released on PyPI a minor version with this PR included: https://pypi.org/project/drf-nested-routers/0.94.0/

alanjds commented:
> so don't release this yet (in case you were planning to). 😄

Oops. Looks like I will need to release a patch very soon.
